Serverless step functions

zirkelc opened the pull request:
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [X] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [X] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [X] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

zirkelc commented:
I submitted a PR https://github.com/microsoft/DefinitelyTyped-tools/pull/732 to allow [@serverless/typescript](https://www.npmjs.com/package/@serverless/typescript) as dependency.

massfords commented on types/serverless-step-functions/index.d.ts:
```diff
@@ -0,0 +1,310 @@
+    [state: string]: Choice | Fail | Map | Task | Parallel | Pass | Wait | Succeed;
+};
+
+type StateBase = {
+    Catch?: Catcher[];
+    Retry?: Retrier[];
+    End?: boolean;
+    InputPath?: string;
+    Next?: string;
+    OutputPath?: string;
+    ResultPath?: string | null;
+    ResultSelector?: { [key: string]: string | { [key: string]: string } };
+    Type: string;
+    Comment?: string;
+};
```
There are fields in this base type that are not permitted on some of the subtypes. For example, a `Choice` cannot have the `End` property.

I suggest removing this type and add the necessary fields where needed in the specific state types

[choice docs](https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-choice-state.html) for reference.

massfords commented on types/serverless-step-functions/index.d.ts:
```diff
@@ -0,0 +1,310 @@
+    Error?: string;
+}
+
+interface Map extends StateBase {
+    Type: 'Map';
+    ItemsPath?: string;
+    Iterator: Definition;
+    ItemSelector?: {
+        [key: string]: string;
+    };
+    MaxConcurrency?: number;
+    Parameters?: {
+        [key: string]: string;
+    };
+}
```
`Map` is missing the `ItemProcessor` field which was added in the Dec 2022 update.

There are some checks in [asl-validator](https://github.com/ChristopheBougere/asl-validator) to reject definitions that have both deprecated and new fields for the same state. i.e. a `Map` that defines both `Parameters` and `ItemSelector.` It would be good to encode this in the TS definition.

ebisbe commented:
@massfords Could we generate the types from the Json schemas? That would make it much safer to update and maintain. Here in the asl-validator I started a conversation https://github.com/ChristopheBougere/asl-validator/discussions/146#discussioncomment-7004484

zirkelc commented:
Hi @massfords and @ebisbe 

I spent some time figuring out how to generate the types from the  JSON schemas of [ASL validator](https://github.com/ChristopheBougere/asl-validator/tree/main/src/schemas), but the json-to-typescript generator endws up failing because of too much recursion and exceeding the stack trace.

So I manually went through all the docs of [AWS StepFunctions ](https://docs.aws.amazon.com/step-functions/latest/dg/concepts-states.html) and checked every possible state to see which fields are allowed. I added the sources to the types as comments. 

The official Amazon State Language specification has a bit more information: https://states-language.net/#state-type-table

Could you guys please give it a thorough look?

Regarding the failing test: it seems like it's an issue in the external package `@serverless/typescript` which in included as dependency in order to use module augmentation. Is it possible to prevent the CI from checking the external packages?

massfords commented on types/serverless-step-functions/types/map.d.ts:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * @deprecated
+     */
+    Paramaters?: JsonObject;
```
This should be `Parameters`

zirkelc commented:
Hi @massfords 

I created an issue on the [@serverless/typescript](https://github.com/serverless/typescript/issues/82) to address the error in the failing test case. Is there something we can do to get this thing forward until the issue is resolved?

massfords commented:
> I created an issue on the [@serverless/typescript](https://github.com/serverless/typescript/issues/82) to address the error in the failing test case. Is there something we can do to get this thing forward until the issue is resolved?

Not sure what there is to do to move this forward. I'm not a contributor to this repo.

zirkelc commented:
Hi @massfords I removed the `@serverless/typescript` dependency to get the tests to succeed. 

When the [PR#82](https://github.com/serverless/typescript/issues/82) on the serverless repo is fixed, I will add the dependency back to the types.

ebisbe commented:
@zirkelc I'll try this afternoon to set up my repo with that one and see if any issues shows up. It's a shame we cannot autogenerate them...

massfords commented on types/serverless-step-functions/types/state.d.ts:
```diff
@@ -0,0 +1,31 @@
+export type IntrinsicFunction = string;
+export type Resource = string | { 'Fn::GetAtt': [string, 'Arn'] } | { 'Fn::Join': [string, Resource[]] };
+
+export type JsonPath = string; // constraint: starts with $.
```
There are two kinds of [paths](https://states-language.net/#path) in ASL. A `Path` and a `ReferencePath`. Both are effectively `string` to Typescript so there's no difference at compile time.

It could be useful from a tooling perspective to model them separately. For example, the schemas in asl-validator use separate names for these to enable validation of the expressions.

Also note that both can start with `$$.` when pulling data from the context

massfords commented on types/serverless-step-functions/types/task.d.ts:
```diff
@@ -0,0 +1,25 @@
+import { IntrinsicFunction, JsonObject, JsonPath, PositiveInteger, Resource } from './state';
+
+// https://docs.aws.amazon.com/step-functions/latest/dg/amazon-states-language-task-state.html#task-state-fields
+export interface Task {
+    Type: 'Task';
+    Comment?: string;
+    InputPath?: JsonPath;
+    OutputPath?: JsonPath;
+    Next?: string;
+    End?: boolean;
+    Resource: Resource;
+    Parameters?: JsonObject;
+    Credentials?: string | IntrinsicFunction | JsonPath;
+    ResultPath?: JsonPath;
+    ResultSelector?: JsonObject;
+    Retry?: Retry[];
+    Catch?: Catch[];
+    TimeoutSeconds?: PositiveInteger;
+    TimeoutSecondsPath?: JsonPath;
+    HeartbeatSeconds?: PositiveInteger;
+    HeartbeatSecondsPath?: JsonPath;
```
This looks good but it allows for invalid combinations of properties that could be enforced by the type definition.

For example:
- a `Task` must have a `Next` or `End` but not both
- a `Task` may have a `TimeoutSeconds` or `TimeoutSecondsPath` but not both
- a `Task` may have a `HeartbeatSeconds` or `HeartbeatSecondsPath` but not both

These rules could be encoded into the type like so: 

```typescript
export type Task = {
    Type: 'Task';
    Comment?: string;
    InputPath?: JsonPath;
    OutputPath?: JsonPath;
    Resource: Resource;
    Parameters?: JsonObject;
    Credentials?: string | IntrinsicFunction | JsonPath;
    ResultPath?: JsonPath;
    ResultSelector?: JsonObject;
    Retry?: Retry[];
    Catch?: Catch[];
} & EndOrNext & TimeoutSecondsOrPath & HeartbeatSecondsOrPath


export type EndOrNext = {
    Next: string;
    End?: never;
} | {
    Next?: never;
    End: boolean;
}

export type TimeoutSecondsOrPath = {
    TimeoutSeconds?: PositiveInteger;
    TimeoutSecondsPath?: never;
} | {
    TimeoutSeconds?: never;
    TimeoutSecondsPath?: JsonPath;
}

export type HeartbeatSecondsOrPath = {
    HeartbeatSeconds?: PositiveInteger;
    HeartbeatSecondsPath?: never;
} | {
    HeartbeatSeconds?: never;
    HeartbeatSecondsPath?: JsonPath;
}

```

ebisbe replied:
Those change look good although I couldn't test them. Also I didn't know that could be done.

massfords reviewed (approved):
The types here look reasonable

I left a couple of comments on some possible changes but nothing urgent.

ebisbe commented:
I've been trying to test it with what they suggest https://github.com/DefinitelyTyped/DefinitelyTyped#testing but I can't... Is there a way to install `@types/serverless-step-functions` from this branch?

zirkelc commented:
@ebisbe no, I think installing from a branch doesn't work in this case. You probably need to opy the `types/serverless-step-functions` folder into your project and then follow the description here: https://github.com/DefinitelyTyped/DefinitelyTyped#adding-tests-to-a-new-package

zirkelc commented:
@massfords I refactored the types to use include the right fields and added a link to the specification for each type.

Please give it another look if I oversaw something.

peterblazejewicz commented on types/serverless-step-functions/index.d.ts:
```diff
@@ -0,0 +1,85 @@
+// Type definitions for serverless-step-functions 3.14
+// Project: https://github.com/serverless-operations/serverless-step-functions
+// Definitions by: Chris Cook <https://github.com/zirkelc>
+//                 Enric Bisbe Gil <https://github.com/ebisbe>
```
this requires explicit ACK from this person, as owners are notified and part of automated communication from DT repo. You can give a credit/kudos instead, when lacking ack

zirkelc replied:
These are my name and @ebisbe - I could add @massfords as well if he's interested?

ebisbe replied:
@peterblazejewicz I'm ok

peterblazejewicz commented on types/serverless-step-functions/index.d.ts:
```diff
@@ -0,0 +1,85 @@
+// Definitions by: Chris Cook <https://github.com/zirkelc>
+//                 Enric Bisbe Gil <https://github.com/ebisbe>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// import { AwsResourceDependsOn, AwsResourceTags } from '@serverless/typescript';
```
```suggestion
```
no dead code

peterblazejewicz commented on types/serverless-step-functions/package.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "private": true,
+    "dependencies": {}
```
why this is empty?

peterblazejewicz commented on types/serverless-step-functions/types/errors.d.ts:
```diff
@@ -0,0 +1,34 @@
+    MaxDelaySeconds?: PositiveInteger;
+    JitterStrategy?: 'FULL' | 'NONE';
+}
+
+export {};
```
```suggestion
```
not required here imo

peterblazejewicz commented on types/serverless-step-functions/types/pass.d.ts:
```diff
@@ -0,0 +1,21 @@
+    ResultPath?: ReferencePath;
+    Parameters?: JsonObject;
+} & EndOrNext;
+
+export {};
```
```suggestion
```
not required imo

peterblazejewicz commented on types/serverless-step-functions/types/state.d.ts:
```diff
@@ -0,0 +1,60 @@
+          Next?: never;
+          End: boolean;
+      };
+
+export {};
```
```suggestion
```
ditto

peterblazejewicz commented on types/serverless-step-functions/types/succeed.d.ts:
```diff
@@ -0,0 +1,17 @@
+    InputPath?: Path;
+    OutputPath?: Path;
+}
+
+export {};
```
```suggestion
```
ditto

peterblazejewicz commented on types/serverless-step-functions/index.d.ts:
```diff
@@ -0,0 +1,85 @@
+}
+
+export = StepFunctions;
+export {};
```
```suggestion
```
and also here

zirkelc commented:
@peterblazejewicz I removed the empty exports. I added them initially to get rid of this linting error:

> ERROR: 36:6   strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/microsoft/DefinitelyTyped-tools/blob/master/packages/dtslint/docs/strict-export-declare-modifiers.md

However, exporting all types does also solve it.

I removed the empty package.json. I added a dependency to `serverless-step-functions` to use module augmentation to extend the existing types. However, the build failed due to a TS error in the original package. I created an [issue](https://github.com/serverless/typescript/issues/82) and referred to it in the types as `TODO` comment.

Let me know what you think!

ebisbe commented on types/serverless-step-functions/types/state.d.ts:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @see https://states-language.net/#ref-paths
+ */
+export type ReferencePath = string;
```
```suggestion
export type ReferencePath = string | null;
```
At least when used in a `ResultPath` pass state that you want to discard the result and pass the original output. If ReferencePath should always be string then we should make that change directly on the pass state

zirkelc replied:
I added `null` to all `InputPath`, `OutputPath` and `ResultPath` according to https://states-language.net/#using-inputpath-parameters-resultselector-resultpath-and-outputpath

ebisbe commented on types/serverless-step-functions/types/state.d.ts:
```diff
@@ -0,0 +1,58 @@
+    [member: string]: JsonValue;
+}
+export interface JsonArray extends Array<JsonValue> {}
+export type JsonValue = JsonPrimitive | JsonObject | JsonArray;
```
Does this accept: 
`{ something: null }`

zirkelc replied:
Yes, `JsonPrimitive` contains `null`

peterblazejewicz reviewed (approved):
LGTM!
@zirkelc thanks!

zirkelc commented:
Ready to merge
